src/main/java/net/earthcomputer/clientcommands/command/FourInARowCommand.java

```diff
@@ -0,0 +1,357 @@
+package net.earthcomputer.clientcommands.command;
+
+import com.mojang.blaze3d.platform.InputConstants;
+import com.mojang.brigadier.CommandDispatcher;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import com.mojang.logging.LogUtils;
+import net.earthcomputer.clientcommands.c2c.C2CPacketHandler;
+import net.earthcomputer.clientcommands.c2c.packets.PutFourInARowPieceC2CPacket;
+import net.earthcomputer.clientcommands.features.TwoPlayerGame;
+import net.earthcomputer.clientcommands.features.TwoPlayerGameType;
+import net.fabricmc.fabric.api.client.command.v2.FabricClientCommandSource;
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.gui.GuiGraphics;
+import net.minecraft.client.gui.screens.Screen;
+import net.minecraft.client.multiplayer.PlayerInfo;
+import net.minecraft.network.chat.ClickEvent;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.HoverEvent;
+import net.minecraft.network.chat.MutableComponent;
+import net.minecraft.resources.ResourceLocation;
+import org.slf4j.Logger;
+
+public class FourInARowCommand {
+    private static final Logger LOGGER = LogUtils.getLogger();
+
+    public static void register(CommandDispatcher<FabricClientCommandSource> dispatcher) {
+        dispatcher.register(TwoPlayerGameType.FOUR_IN_A_ROW_GAME_TYPE.createCommandTree());
+    }
+
+    public static void onPutFourInARowPieceC2CPacket(PutFourInARowPieceC2CPacket packet) {
+        String sender = packet.sender();
+        FourInARowGame game = TwoPlayerGameType.FOUR_IN_A_ROW_GAME_TYPE.getActiveGame(sender);
+        if (game == null) {
+            return;
+        }
+        game.onMove(packet.x());
+    }
+
+    public static class FourInARowGame extends TwoPlayerGame<FourInARowGameScreen> {
+        public static final int WIDTH = 7;
+        public static final int HEIGHT = 6;
+
+        public final PlayerInfo opponent;
+        public final byte yourPiece;
+        public byte activePiece;
+        public final byte[][] board;
+        /**
+         * 0 for undecided
+         * 1 for red
+         * 2 for yellow
+         * 3 for draw
+        */
+        public byte winner;
+
+        public FourInARowGame(PlayerInfo opponent, byte yourPiece) {
+            this.opponent = opponent;
+            this.yourPiece = yourPiece;
+            this.activePiece = Piece.RED;
+            this.board = new byte[WIDTH][HEIGHT];
+            this.winner = 0;
+        }
+
+        public void onMove(int x) {
```

Comment:
onMove is called by both interactions since a lot of the logic can be copied between your moves and their moves. The function itself handles for when a move is done, the logic to check if you can make that move or not right now is located elsewhere, are you asking for me to merge them?

Reply:
The same is true for `putMark`, it is also used by yourself and the opponent. As far as I can tell, when the opponent makes a move, it is not checked whether that move is valid. The move packet is handled in `onPutFourInARowPieceC2CPacket`, whereafter `onMove` is called. Inside `onMove` there are no checks for whether the move is valid. (We cannot trust the other client.)

If you do it tic-tac-toe style, you include the opponent's `piece`s in the call to `onMove`, and then _I think_ you can concisely check if `(this.yourPiece ^ piece ^ this.activePiece) > 0` and make the move.